libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/ChargingStationsMetadataUpdater.kt

```diff
@@ -0,0 +1,134 @@
+    serverProvidedStationsIds: Set<String>
+): DirectionsResponse {
+    val updatedRoutes = response.routes().map { route ->
+        val updatedWaypoints =
+            (route.waypoints() ?: emptyList()).mapIndexed { waypointIndex, waypoint ->
+                val waypointUnrecognizedProperties = waypoint.unrecognizedJsonProperties
+                val waypointMetadata = waypointUnrecognizedProperties?.get("metadata")
+                    ?.asJsonObject
+                val stationId = waypointMetadata
+                    ?.get("station_id")
+                    ?.asString
+                if (stationId != null && serverProvidedStationsIds.contains(stationId)) {
+                    val updatedMetadata = waypointMetadata.deepCopy()
+                    updatedMetadata.remove("type")
+                    updatedMetadata.remove("was_requested_as_user_provided")
+                    updatedMetadata.add("type", JsonPrimitive("charging-station"))
+                    updatedMetadata.add("was_requested_as_user_provided", JsonPrimitive(true))
+                    val updatedUnrecognizedProperties = waypointUnrecognizedProperties
+                        .toMutableMap()
+                    updatedUnrecognizedProperties["metadata"] = updatedMetadata
+                    waypoint.toBuilder()
+                        .unrecognizedJsonProperties(updatedUnrecognizedProperties)
+                        .build()
+                } else {
+                    waypoint
+                }
+            }
+        route.toBuilder().waypoints(updatedWaypoints).build()
+    }
+    val updatedResponse = response.toBuilder().routes(updatedRoutes).build()
+    return updatedResponse
+}
+
+private fun updateChargingStationsMetadataBasedOnRequestUrl(
+    rerouteRouteOptions: RouteOptions,
+    newPrimaryRoute: NavigationRoute
+): DirectionsResponse {
+    val chargingStationsPowers = rerouteRouteOptions
+        .getUnrecognisedSemicolonSeparatedParameter("waypoints.charging_station_power")
+    val chargingStationsIds = rerouteRouteOptions
+        .getUnrecognisedSemicolonSeparatedParameter("waypoints.charging_station_id")
+    val chargingStationsCurrentTypes = rerouteRouteOptions
+        .getUnrecognisedSemicolonSeparatedParameter("waypoints.charging_station_current_type")
+    val response = newPrimaryRoute.directionsResponse
+    val updatedRoutes = response.routes().map { route ->
+        val updatedWaypoints =
+            (route.waypoints() ?: emptyList()).mapIndexed { waypointIndex, waypoint ->
+                val chargingStationId = chargingStationsIds?.get(waypointIndex)
+                val chargingStationsPower = chargingStationsPowers?.get(waypointIndex)
+                    ?.toIntOrNull()
+                    ?.let { it / 1000 }
+                val chargingStationsCurrentType = chargingStationsCurrentTypes?.get(waypointIndex)
+                waypoint.toBuilder()
+                    .unrecognizedJsonProperties(
+                        mapOf(
```

Comment:
I think you should also add "type": "user-provided-charging-station" here.

Reply:
done